Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs

```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+
+	/// <summary>
+	/// Triggers the <see cref="OnShown"/> event. Allows interception of the event in derived components.
+	/// </summary>
+	protected virtual Task InvokeOnShownAsync(string elementId) => OnShown.InvokeAsync(elementId);
+
+	private IJSObjectReference jsModule;
+	private readonly DotNetObjectReference<HxMultiSelectInternal<TValue, TItem>> dotnetObjectReference;
+	private ElementReference elementReference;
+	private ElementReference filterInputReference;
+	private bool isShown;
+	private string filterText = string.Empty;
+	private bool disposed;
 
-	private async Task HandleItemSelectionChangedAsync(bool newChecked, TItem item)
+	public HxMultiSelectInternal()
+	{
+		dotnetObjectReference = DotNetObjectReference.Create(this);
+	}
+
+	/// <inheritdoc cref="ComponentBase.OnAfterRenderAsync(bool)" />
+	protected override async Task OnAfterRenderAsync(bool firstRender)
+	{
+		await base.OnAfterRenderAsync(firstRender);
+
+		if (firstRender)
+		{
+			await EnsureJsModuleAsync();
+			if (disposed)
+			{
+				return;
+			}
+
+			await jsModule.InvokeVoidAsync("initialize", elementReference, dotnetObjectReference, false);
+		}
+	}
+
+	private async Task EnsureJsModuleAsync()
+	{
+		jsModule ??= await JSRuntime.ImportHavitBlazorBootstrapModuleAsync(nameof(HxMultiSelect));
+	}
+
+	private async Task HandleItemSelectionChangedAsync(bool newChecked, TItem item, bool triggerSelectAllEvent = true)
 	{
 		await ItemSelectionChanged.InvokeAsync(new SelectionChangedArgs
 		{
 			Checked = newChecked,
 			Item = item
 		});
+
+		if (triggerSelectAllEvent)
+		{
+			await ChangeSelectAllAsync(false);
+		}
+	}
+
+	private bool selectAll;
+
+	private async Task HandleSelectAllClickedAsync()
+	{
+		var filteredItems = GetFilteredItems();
+
+		// If all items are already selected then they should be deselected, otherwise only records that aren't selected should be
+		if (selectAll)
+		{
+			foreach (var item in filteredItems)
+			{
+				// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+				await HandleItemSelectionChangedAsync(false, item, false);
+			}
+		}
+		else
+		{
+			foreach (var item in filteredItems)
+			{
+				var value = SelectorHelpers.GetValue<TItem, TValue>(ValueSelector, item);
+				var itemSelected = DoSelectedValuesContainValue(value);
+
+				// If the item is already selected we don't need to reselect it
+				if (!itemSelected)
+				{
+					// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+					await HandleItemSelectionChangedAsync(!itemSelected, item, false);
+				}
+			}
+		}
+
+		await ChangeSelectAllAsync(!selectAll);
+	}
+
+	private bool DoSelectedValuesContainValue(TValue value)
+	{
+		return SelectedValues?.Contains(value) ?? false;
+	}
+
+	private Task ChangeSelectAllAsync(bool selectAll)
+	{
+		this.selectAll = selectAll;
+		return SelectAllChanged.InvokeAsync(selectAll);
+	}
```

Comment:
(I think we should not expose the SelectAllChanged event (EventCallback) as parameter et al.)

In case we keep the `SelectAllChanged` parameter, we should check if the `selectAll` value changed and invoke the `EventCallback` only if the value really changed. (Now it gets triggered from `HandleItemSelectionChangedAsync` with every single selection of individual item in the list.)